Fixes #17282: Documentation for group properties (#712)
Normation/rudder-doc: src/reference/modules/reference/pages/variables.adoc

"example":{ 
  "group-added"    : "group-added"
, "replaced-string": "group"
, "replaced-array" : {"replaced":"group"}
, "replaced-object": ["group"]
, "append-array"   : [3,4]
  "merge-object"   : {
      "group-added"    : "group-added"
    , "parent-replaced": "group"
  }
}

// on nodes belonging to previous group, which doesn't have "example"
// property defined (so no more overriding at that level):
"example":{ 
  "parent-kept"    : "parent-kept"
, "group-added"    : "group-added"
, "replaced-string": "group"
, "replaced-array" : {"replaced":"group"}
, "replaced-object": ["group"]
, "append-array"   : [1,2,3,4]
, "merge-object"   : {
      "parent-kept"    : "parent-kept"
    , "group-added"    : "group-added"
    , "parent-replaced": "group"
  }
}
----

==== Global parameter

Using this, you can specify common file headers (this is the default parameter, "rudder_file_edit_header"), common DNS or domain names, backup servers,
site-specific elements...

Rudder provides a simple way to add common and reusable variables in either plain directives, or techniques created using the technique editor: the parameters.

image::rudder-parameters.png[Parameters]

==== Group properties

> Group properties can be managed by using https://docs.rudder.io/api/#operation/updateGroup[goup's API] or they can be found in the *properties* tab of each group in Rudder UI.

Reviewer: group

We present afterward a subset of interesting cases, but there is others that you can check out in https://docs.rudder.io/reference/latest/reference/generic_methods.html#_variable[generic method documentation].

====== Variable override by order of rules and directives

Variables defined with following generic methods or corresponding techniques can be overridden by other variables with same name defined in other rules or directives. 

The exact ordering and rules are explain in the https://docs.rudder.io/reference/latest/usage/advanced_configuration_management.html#_sorting_directives_based_on_the_same_technique[ordering directive] chapter.

====== Variable from files

Rudder can natively parse JSON, YAML and CSV files to transform them into JSON values.

====== Variable from `Augeas`

http://augeas.net[Augeas] is standard tool for reading and editing configuration files by providing a tree-representation of their native format. It's really useful to clean the mess in that domain. With it, you can easily and safely transform almost any configuration file into a Rudder JSON variable. 

> ====== Variable from `OSQuery`

Reviewer: ```suggestion
====== Variable from `osquery`
```

==== Node properties

Node properties can be managed by using https://docs.rudder.io/api/#operation/updateNode[node's API] or they can be found in the *properties* tab of each node in Rudder UI.

image::node-properties.png[Node properties in UI]

==== Configuration data example

This section provides a full example of node property definition for the following use case: 

image::node-properties-hierarchy.png[Hierarchical property definition and overriding]

==== Visualizing property inheritance

It is important to be able to understand where a property is defined and how it is overriden to be able to manage it efficiently (and more importantly, to be able to undertand why it's not what it is supposed to be). 

Rudder allows to see a property lineage in group and node property tab: inherited (even if latter overrided) properties get a dedicated tag and going on that tag show its full definition and overrides. 

image::view-property-inheritance.jpg[Viewing property overrides]

==== Syntax

===== Property syntax

[cols="3", options="header"]
|===
|Use case
|name=datacenter, value="Paris"
|name=datacenter, value={"dns": "1.1.1.1"}

> |Technique CFEngine

Reviewer: Classic technique

|${node.properties[datacenter]}
|${node.properties[datacenter][dns]}

> |Technique DSC

Reviewer: DSC technique

Informations about a node's policy server:

[cols="2", options="header"]
|===
|Variable
|Description

|${rudder.node.policyserver.id}
|The Rudder generated id of the Policy Server

|${rudder.node.policyserver.hostname}
|The hostname of the Policy Server

|${rudder.node.policyserver.admin}
|The administrator login of the Policy Server

|===

====== Node-level system properties and constant

[WARNING]

====

These variables are not available on Windows nodes, but 
only on with the classic Linux/AIX agent.

====

NOTE: These properties are evaluated on the node at run time.

====== Syntax

[cols="3", options="header"]
|===
|Use case
|name=sys.host, value="host.local.name"
|No JSON like values

|Technique CFEngine
|${sys.host}
|N/A

|Technique DSC
|N/A
|N/A

|Directive
|${sys.host}
|N/A

|Mustache
|{{{vars.sys.host}}}
|N/A

|Jinja2
|{{vars.sys.host}}
|N/A

|===

====== System property

[cols="2", options="header"]
|===
|Name
|Description

|sys.arch
|Kernel short architecture

|sys.fqhost
|Fully qualified hostname, as seen in Rudder

|sys.uqhost
|Unqualified hostname

|sys.host
|Node's hostname (according to the kernel)

|sys.domain
|Node's domain as discovered by the agent
|===

> There are also more variables available, all documented in https://docs.cfengine.com/docs/3.10/reference-special-variables-sys.html[this page].

Reviewer: ```suggestion
There are also more variables available, all documented in https://docs.cfengine.com/docs/3.15/reference-special-variables-sys.html[this page].
```

==== Node properties

Node properties can be managed by using https://docs.rudder.io/api/#operation/updateNode[node's API] or they can be found in the *properties* tab of each node in Rudder UI.

image::node-properties.png[Node properties in UI]

==== Configuration data example

This section provides a full example of node property definition for the following use case: 

image::node-properties-hierarchy.png[Hierarchical property definition and overriding]

==== Visualizing property inheritance

It is important to be able to understand where a property is defined and how it is overriden to be able to manage it efficiently (and more importantly, to be able to undertand why it's not what it is supposed to be). 

Rudder allows to see a property lineage in group and node property tab: inherited (even if latter overrided) properties get a dedicated tag and going on that tag show its full definition and overrides. 

image::view-property-inheritance.jpg[Viewing property overrides]

==== Syntax

===== Property syntax

[cols="3", options="header"]
|===
|Use case
|name=datacenter, value="Paris"
|name=datacenter, value={"dns": "1.1.1.1"}

|Technique CFEngine

> |${node.properties[datacenter]}

Reviewer: ```suggestion
| `${node.properties[datacenter]}`
```

= Variables

== Generalities

Rudder has a notion of `variables`. A `variable` is something with a `name` that references a `value`, and that `value` can be accessed with appropriate syntax around its `name`. We call accessing the value `variable expansion` in honor to shell, guarding totem of devops glue. 

> In Rudder, `variable name` are namespaced with a dotted convention. `Values` can be maonly of two types: `string` and `json`, ie dictionnary-like recursice structures. Rudder only supports `UTF-8` for `varibles` `names` and `values`.

Reviewer: ```suggestion
In Rudder, `variable name` are namespaced with a dotted convention. `Values` can be mainly of two types: `string` and `json`, ie dictionnary-like recursive structures. Rudder only supports `UTF-8` for `variables` `names` and `values`.
```

Reviewer: Actually I think variable names need to be ASCII

Author: the need for cfengine, but they are escaped by agent during generation. Is there an other reason to do so?

Reviewer: They are put as is in properties.json, but when trying to use a variable name containing an emoji in a directive parameter I get:

> ⇨ Accumulated: Unexpected: Error with parameters expansion for node 'server.rudder.local' (root); cause was: Accumulated: When processing directive 'File content'; cause was: Accumulated: On variable 'GENERIC_FILE_CONTENT_PAYLOAD':; cause was: Accumulated: Unexpected: Error when parsing value "${node.properties[emo👋ji]}", error message is:

We can keep UTF-8, but what I fear is that a lot of things will break in unexpected ways.

Author: OK, what shoud we authorize? `[\-_0-9a-zA-Z]`? `~` ? All ascii without control characters?

Under that broad notion of `variable`, we need to take appart the two main action around them:

- how one defines them? 
- and how one uses them?

In each case, we have subacase directed by use. 

Variables definition covers two big aspects of knowledge:

> - configuration data, known by humans when policies are defined. That knowledge need to be centralized, categorized, hierarchized and retrieved based on what node

Reviewer: ```suggestion
- configuration data, known by humans when policies are defined. That knowledge needs to be centralized, categorized, hierarchized and retrieved based on what node
```

 policies will be used, 

> - dynamic or contextualized information about infrastructure provided by environment tooling, be it Rudder itsefl, facts about environment, or resources available only when the policy is checked. 

Reviewer: ```suggestion
- dynamic or contextualized information about infrastructure provided by environment tooling, be it Rudder itself, facts about environment, or resources available only when the policy is checked. 
```
